### PR TITLE
feat: UI Component Range Calendar

### PR DESCRIPTION
**Pull request by dgodinez-dh**
closes https://github.com/deephaven/deephaven-plugins/issues/904
closes https://github.com/deephaven/deephaven-plugins/issues/854

### CODE REVIEW

**Review by dsmmcken — CHANGES_REQUESTED** [review]
docs/sidebar.json should be updated to link to relevant page (alphabetical order).

I think we've missed that for some of your previous components as well though.

File: plugins/ui/DESIGN.md [hunk 1]
```diff
@@ -2070,6 +2069,139 @@ ui.picker(
...
+    on_change: Callable[[{ "start": Date, "end": Date }], None] | None = None,
+    **props: Any
+) -> RangeCalendarElement
+```
+
+###### Parameters
+
+| Parameter               | Type                                                       | Description                                                                                        |
+| ----------------------- | ---------------------------------------------------------- | -------------------------------------------------------------------------------------------------- |
+| `value`                 | `{ "start": Date, "end": Date } \| None`                   | The current value (controlled).                                                                    |
```
Comment (jnumainville) [inline plugins/ui/DESIGN.md:2124]: This typing is a bit odd to me - any reason we don't just use the `DateRange` type here?

**Review by jnumainville — CHANGES_REQUESTED** [review]

File: plugins/ui/DESIGN.md [hunk 2]
```diff
@@ -2070,6 +2069,139 @@ ui.picker(
...
+instant_end = to_j_instant("2022-01-05T00:00:00 ET")
+local_start = to_j_local_date("2024-05-06")
+local_end = to_j_local_date("2024-05-10")
+
+# simple range calendar that takes a range and is uncontrolled
+range_calendar1 = ui.range_calendar(
+    default_value={"start": local_start, "end": local_end}
+)
+
+# simple range calendar that takes list view items directly and is controlled
```
Comment (bmingles) [inline plugins/ui/DESIGN.md:2150]: Did you mean to say "list view items" here, or is this a copy / paste artifact?

File: plugins/ui/docs/sidebar.json [hunk 3]
```diff
@@ -78,9 +78,13 @@
...
           },
           {
-            "label": "range_slider",
+            "label": "range_calendar",
             "path": "components/range_slider.md"
           },
+          {
+            "label": "range_slider",
+            "path": "components/range_calendar.md"
+          },
```
Comment (dsmmcken) [inline plugins/ui/docs/sidebar.json:87]: that doesn't look right, label and paths are swapped

File: plugins/ui/DESIGN.md (hunk 1 above)
Comment (dgodinez-dh) [reply plugins/ui/DESIGN.md]: Updated

File: plugins/ui/DESIGN.md (hunk 2 above)
Comment (dgodinez-dh) [reply plugins/ui/DESIGN.md]: Yes. That was a copy / paste artifact. Updated.

File: plugins/ui/docs/sidebar.json (hunk 3 above)
Comment (dgodinez-dh) [reply plugins/ui/docs/sidebar.json]: Fixed

File: plugins/ui/src/js/src/elements/RangeCalendar.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  props: SerializedRangeCalendarProps<DHCRangeCalendarProps<DateValue>>
+): JSX.Element {
+  const {
+    defaultValue = null,
+    value: propValue,
+    onChange: propOnChange = EMPTY_FUNCTION,
+    ...otherProps
+  } = useRangeCalendarProps(props);
+
+  const [value, onChange] = useDebouncedOnChange<RangeValue<DateValue> | null>(
```
Comment (bmingles) [inline plugins/ui/src/js/src/elements/RangeCalendar.tsx:26]: This seems to make the component always controlled. If the `default_value` prop is updated, it changes the value.

e.g.

```python
from deephaven import ui


@ui.component
def example():
    default_value, set_default_value = ui.use_state({"start": "2020-02-03", "end": "2020-02-08"})

    return ui.range_calendar(
        aria_label="Date range (uncontrolled)",
        default_value=default_value,
    ), ui.button('Update default', on_press=lambda _: set_default_value({"start": "2020-02-09", "end": "2020-02-20"}))


my_example = example()
```

File: plugins/ui/docs/components/range_calendar.md [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+
+my_range_calendar_is_read_only_example = ui.range_calendar(
+    is_read_only=True,
+)
+```
+
+## Visible Months
+
+By default, the range calendar displays a single month. The `visible_Months` prop allows displaying up to 3 months at a time.
```
Comment (margaretkennedy) [inline plugins/ui/docs/components/range_calendar.md:195]:
```suggestion
By default, the range calendar displays a single month. The `visible_months` prop allows displaying up to 3 months at a time.
```

File: plugins/ui/src/js/src/elements/utils/DateTimeUtils.test.ts [hunk 6]
```diff
@@ -201,3 +207,22 @@ describe('parseCalendarValue', () => {
...
+
+describe('dateValuetoIsoString', () => {
+  it('handles a CalendarDate', () => {
+    const date = new CalendarDate(2021, 3, 4);
+    expect(dateValuetoIsoString(date)).toEqual('2021-03-04');
+  });
+
+  it('handles a CalendarDateTime', () => {
+    const date = new CalendarDateTime(2021, 3, 4, 5, 6, 7);
+    expect(dateValuetoIsoString(date)).toEqual('2021-03-04T05:06:07Z');
```
Comment (bmingles) [inline plugins/ui/src/js/src/elements/utils/DateTimeUtils.test.ts:219]: I like the sequential numbering. Makes it easy to see the relationship of the input / output.

**Review by bmingles — CHANGES_REQUESTED** [review]
All looks good except there may be a bug with the `default_value` being controlled instead of uncontrolled.

**Review by bmingles — DISMISSED** [review]
LGTM

File: plugins/ui/src/js/src/elements/RangeCalendar.tsx (hunk 4 above)
Comment (dgodinez-dh) [reply plugins/ui/src/js/src/elements/RangeCalendar.tsx]: Updated

**Review by jnumainville — APPROVED** [review]

**Review by dsmmcken — APPROVED** [review]
approving docs.

**Pull request merged**